### ISSUE TITLE
[飞桨复现论文]batch=1时出现nan

### ISSUE THREAD

**Issue by zhoucz97** [PaddlePaddle/Paddle#36665]
情况：数据量按照batch_size切分后还剩余的一个数据，或者直接把batch_size设为1

问题：在把这一个数据送入模型后，得到loss，并计算梯度反向传播后，整个模型出现问题。具体问题是模型权重全部变为nan，后续梯度也都变为nan。

**在送入模型的数据量不是1的情况下不会出现此问题。
即使数据量为1，但在CPU上也不会出现此问题。**
**只有在送入模型的数据量为1且在GPU上，才会出现此问题。**

**Comment by yghstill** [comment]
@zhoucz97 Paddle版本是多少？是GPU单卡训练出现这个问题吗？是什么类型的模型，可否提供下最小可复现代码，以便我们定位问题并解决?

**Comment by zhoucz97** [comment]
ai studio地址和一些说明放在文件中了，谢谢回复~~




周长智
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/Paddle"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年10月23日(星期六) 下午5:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [PaddlePaddle/Paddle] [飞桨复现论文]batch=1时出现nan (Issue #36665)





 
@zhoucz97 Paddle版本是多少？是GPU单卡训练出现这个问题吗？是什么类型的模型，可否提供下最小可复现代码，以便我们定位问题并解决?
 
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by yghstill** [comment]
@zhoucz97 issue里好像看不到链接和文件，或者将ai studio链接重发下？

**Comment by zhoucz97** [comment]
> @zhoucz97 issue里好像看不到链接和文件，或者将ai studio链接重发下？

在ai studio上运行的
paddle.__version__=2.1.2
GPU单卡
Splinter 问答BERT类模型
论文：https://paperswithcode.com/paper/few-shot-question-answering-by-pretraining


我的ai studio项目链接如下：
项目“splinter-paddle”共享链接(有效期三天)：https://aistudio.baidu.com/studio/project/partial/verify/2503997/e225be8238c14b97b9ed123dfbc364cb


您只需关注splinter-paddle/finetuning中的train.py和run.py文件

2503997.ipynb可直接运行项目，因为是few-shot，训练速度很快，一两分钟就可以训一次。


train.py
167行和169行，有两次梯度输出

modeling.py文件
第101行会输出两个loss中的一个



这个文件"splinter-paddle/mrqa-few-shot/squad/squad-train-seed-42-num-examples-128_qass.jsonl"经过数据处理后会产生133个数据，batch_size=12的话正好余一个，因此会出问题

您也可以改成其他数据集（只需改数据文件名中的128），或者设成其他batch_size，只要最后余一个数据或者直接把batch_size设为1，就会出问题。

我们也在找原因，还不知道是我们代码有问题还是其他地方有问题。但经过排查，我们给模型的输入维度每次都是一样的，都是(batch_size, seq_len, dim)，输出维度也一样，模型参数和原论文也一样。


此外上述的问题在本人电脑上用paddle cpu版本运行，没有问题，所以有点懵。

**Comment by yghstill** [comment]
@zhoucz97 小样本场景下出 nan 的频率会相对较高，小样本场景对超参数很敏感，可以更改一下其它超参数看看是否会出 nan，例如 lr、max_seq_len 等
